persistent_volumes_test.go

```diff
@@ -0,0 +1,410 @@
+			continue
+		}
+		// found a bounded PV
+		p := pv.Spec.Capacity[api.ResourceStorage]
+		pvCap := p.Value()
+		p = resource.MustParse(strconv.Itoa(maxPVs/2) + "G")
+		pvcCap := p.Value()
+		if pv.Spec.ClaimRef.Name != pvc.Name || pvCap != pvcCap {
+			t.Fatalf("Bind mismatch! Expected %s capacity %v but got %s capacity %v", pvc.Name, resource.MustParse(strconv.Itoa(50)+"G"),
+				pv.Spec.ClaimRef.Name, pv.Spec.Capacity[api.ResourceStorage])
+		}
+		t.Logf("claim bounded to %s capacity %v", pv.Name, pv.Spec.Capacity[api.ResourceStorage])
+		bound += 1
+	}
+
+	if bound != 1 {
+		t.Fatalf("Only 1 PV should be bound but got %d", bound)
+	}
+
+	// deleting a claim releases the volume
+	if err := testClient.PersistentVolumeClaims(api.NamespaceDefault).Delete(pvc.Name, nil); err != nil {
+		t.Errorf("error deleting claim %s", pvc.Name)
+	}
+
+	waitForAnyPersistentVolumePhase(w, api.VolumeReleased)
+
+	deleteAllEtcdKeys()
+}
+
+// TestPersistentVolumeMultiPVsPVCs tests binding of 100 PVC to 100 PVs.
+func TestPersistentVolumeMultiPVsPVCs(t *testing.T) {
+	_, s := framework.RunAMaster(t)
+	defer s.Close()
+
+	deleteAllEtcdKeys()
+	testClient, _, binder := createClients(s)
+	binder.Run()
+	defer binder.Stop()
+
+	maxObjects := 1000
+	pvs := make([]*api.PersistentVolume, maxObjects)
+	pvcs := make([]*api.PersistentVolumeClaim, maxObjects)
+	for i := 0; i < maxObjects; i++ {
+		// This PV will be claimed, released, and deleted
+		pvs[i] = createPV("pv-"+strconv.Itoa(i), "/tmp/foo"+strconv.Itoa(i), "1G",
+			[]api.PersistentVolumeAccessMode{api.ReadWriteOnce}, api.PersistentVolumeReclaimDelete)
+		pvcs[i] = createPVC("pvc-"+strconv.Itoa(i), "1G", []api.PersistentVolumeAccessMode{api.ReadWriteOnce})
+	}
+
+	w, _ := testClient.PersistentVolumes().Watch(api.ListOptions{})
+	defer w.Stop()
+
+	for i := 0; i < maxObjects; i++ {
+		_, _ = testClient.PersistentVolumes().Create(pvs[i])
+		_, _ = testClient.PersistentVolumeClaims(api.NamespaceDefault).Create(pvcs[i])
+	}
+
+	// wait until the binder pairs all volumes
+	for i := 0; i < maxObjects; i++ {
+		waitForPersistentVolumePhase(testClient, pvs[i].Name, w, api.VolumeBound)
+		glog.V(1).Infof("Volume %q bound", pvs[i].Name)
+	}
+
+	// check that everything is bound to something
+	for i := 0; i < maxObjects; i++ {
+		pv, err := testClient.PersistentVolumes().Get(pvs[i].Name)
+		if err != nil {
+			t.Fatalf("Unexpected error getting pv: %v", err)
+		}
+		if pv.Spec.ClaimRef == nil {
+			t.Fatalf("PV %q is not bound", pv.Name)
+		}
+
+		pvc, err := testClient.PersistentVolumeClaims(api.NamespaceDefault).Get(pvcs[i].Name)
+		if err != nil {
+			t.Fatalf("Unexpected error getting pvc: %v", err)
+		}
+		if pvc.Spec.VolumeName == "" {
+			t.Fatalf("PVC %q is not bound", pvc.Name)
+		}
+	}
+	deleteAllEtcdKeys()
+}
+
+// TestPersistentVolumeMultiPVsDiffAccessModes tests binding of one PVC to two
+// PVs with different access modes.
+func TestPersistentVolumeMultiPVsDiffAccessModes(t *testing.T) {
```

Comment:
This is basically a check on function that dictates the traversal order of PVs to bind a claim to, yes?